### PR TITLE
[Master] Fix inconsistent attempt to expose non-public type presents in union and object types

### LINKED ISSUE #39180: [Bug]: Inconsistent attempt to expose non-public error
### Description

Program 1 generates an ``attempt to expose non-public symbol 'Employee' `` error, But program 2 is not. Observed this in SL Update 3. 

Program 1:

```
type Employee record {|
|};

client class RainierClient {
    resource function 'select employees() returns stream<Employee, error?> {
    }
}
```

Program 2: 

```
type Employee record {|
|};

client class RainierClient {
    resource function 'select employees() returns stream<Employee, error?>? {
        return;
    }
}
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes #39180

## Approach
Check the `attempt to expose non-public type` in union type and object types

## Samples
```
type Employee1 record {|
|};

type Employee2 record {|
    int x;
|};

public Employee1|Employee2 emp1 = {};
```
This will give the warning `attempt to expose non-public type Employee1`, `attempt to expose non-public type Employee2`

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 1]
```diff
@@ -1686,7 +1686,15 @@ private void checkForExportableType(BTypeSymbol symbol, Location pos, HashSet<BT
...
                 return;
-            // TODO : Add support for other types. such as union and objects
+            case TypeTags.UNION:
+                ((BUnionType) symbolType).getMemberTypes().forEach(t -> checkForExportableType(t.tsymbol, pos,
+                        visitedSymbols));
+                return;
+            case TypeTags.OBJECT:
+                BObjectType objectType = (BObjectType) symbolType;
+                objectType.fields.values().forEach(f -> checkForExportableType(f.type.tsymbol, pos,
+                        visitedSymbols));
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:1696]:
```suggestion
                ((BObjectType) symbolType).fields.values().forEach(f -> checkForExportableType(f.type.tsymbol, pos,
                        visitedSymbols));
```

File: tests/jballerina-unit-test/src/test/resources/test-src/access/private_access_negative.bal [hunk 2]
```diff
@@ -99,7 +108,8 @@ function test3() returns string {
 
     return m.s;
 }
-// TODO: Fix me. This is a bug.
+
+
```
Comment (LakshanWeerasinghe) [inline tests/jballerina-unit-test/src/test/resources/test-src/access/private_access_negative.bal:112]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/access/private_access_negative.bal [hunk 3]
```diff
@@ -87,7 +95,8 @@ public function test2() returns object {
 
     return m;
 }
-// TODO: Fix me. This is a bug.
+
+
```
Comment (LakshanWeerasinghe) [inline tests/jballerina-unit-test/src/test/resources/test-src/access/private_access_negative.bal:99]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectWithPrivateFieldsNegativeTest.java [hunk 4]
```diff
@@ -68,19 +68,23 @@ public void testPrivateObjAccess1SemanticsNegative() {
...
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'ChildFoo'", 5, 5);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 34, 45);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 38, 1);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 42, 1);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 42, 73);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 4, 5);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 6, 5);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 16, 5);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 59, 38);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 66, 44);
```
Comment (ushirask) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectWithPrivateFieldsNegativeTest.java:87]: We can add `Assert.assertEquals(compileResult.getWarnCount(), i)` here similar to  https://github.com/ballerina-platform/ballerina-lang/pull/39970/files#diff-4f8066db20149aa8f27874addc97290589b27889ff3d05ff4ee6ba9b2095aa9fL73 so we won't have to change the number of warnings when updating tests.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectWithPrivateFieldsNegativeTest.java [hunk 5]
```diff
@@ -113,7 +117,11 @@ public void testPrivateObjAccess2SemanticsNegative() {
         BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 38, 1);
         BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 42, 1);
         BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'PrivatePerson'", 42, 73);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 4, 5);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 6, 5);
         BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 16, 5);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 59, 38);
+        BAssertUtil.validateWarning(compileResult, i++, expectedWarningMsg + "'FooFamily'", 66, 44);
```
Comment (ushirask) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectWithPrivateFieldsNegativeTest.java:124]: https://github.com/ballerina-platform/ballerina-lang/pull/39970/files#r1147324621

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectWithPrivateFieldsNegativeTest.java (hunk 4 above)
Comment (prakanth97) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectWithPrivateFieldsNegativeTest.java:85]: Updated ad671ef6b77b2f6f41f3e3d021ea829778d7819c

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 6]
```diff
@@ -1690,7 +1691,18 @@ private void checkForExportableType(BTypeSymbol symbol, Location pos, HashSet<BT
                 symbolType = Types.getReferredType(symbolType);
                 checkForExportableType(symbolType.tsymbol, pos, visitedSymbols);
                 return;
-            // TODO : Add support for other types. such as union and objects
+            case TypeTags.UNION:
+                ((BUnionType) symbolType).getMemberTypes().forEach(t -> checkForExportableType(t.tsymbol, pos,
+                        visitedSymbols));
+                return;
+            case TypeTags.OBJECT:
+                ((BObjectType) symbolType).fields.values().forEach(f -> checkForExportableType(f.type.tsymbol, pos,
```
Comment (ushirask) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:1699]: I think the logic here should be similar to what we have done with record types

Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Updated [9cffdc1](https://github.com/ballerina-platform/ballerina-lang/pull/39970/commits/9cffdc14e7f69d756b3454acb42e91d924499f83)

**Comment by prakanth97** [comment]
deferred

**Pull request closed without merging**